### PR TITLE
chore: Clarify resolution detail reason.

### LINKED ISSUE #236: Clarification of reason for evaluation detail and evaluation.
In the specification it is detailed that the `reason` is an optional string, and there are a set of pre-defined values, but I do not think it is clear that free-form values are intended to be supported as well. This has resulted in some SDK development which restricts the reason to an enumeration and prevents conveying reasons more specific to the vendor.

A current example of this in the python SDK: https://github.com/open-feature/python-sdk/blob/d1f27e3278964c2d81b780db97deace87d02cae1/openfeature/flag_evaluation.py#L21

(A different issue will be opened for that specific instance.)

### PR DESCRIPTION
**Pull request by kinyoklion**
## This PR

Attempt to clarify the `reason` field of the resolution details.

### Related Issues

Fixes #236

### CODE REVIEW

File: specification/types.md [hunk 1]
```diff
@@ -72,6 +72,28 @@ A set of pre-defined reasons is enumerated below:
 | STALE           | The resolved value is non-authoritative or possibly out of date                                                                  |
 | ERROR           | The resolved value was the result of an error.                                                                                   |
 
+> [!NOTE]
+> The `reason` should not be limited to the reasons enumerated above. It can be any of the pre-defined reasons, or
+> any string value. Some type systems have features which can increase the ergonomics of `reason`, for instance a union
+of pre-defined types with a string, or a rust-style enumeration which allows for enumerated values to have associated
+content.
+
+```rust
```
Comment (kinyoklion) [inline specification/types.md:81]: I am not sure if an example here is overkill.

Comment (kinyoklion) [reply specification/types.md:81]: The solution to this varies widely language to language, so maybe just pointing it out is enough. 

The js-sdk:
```typescript
export type ResolutionReason = keyof typeof StandardResolutionReasons | (string & Record<never, never>);
```

The solution added to the python SDK.
```python
reason: typing.Optional[typing.Union[str, Reason]] = None
```

File: specification/types.md [hunk 2]
```diff
@@ -72,6 +72,28 @@ A set of pre-defined reasons is enumerated below:
...
+enum Reason {
+    Static,
+    Default,
+    TargetingMatch,
+    Split,
+    Cached,
+    Unknown,
+    Stale,
+    Error,
+    Other(String)
```
Comment (kinyoklion) [inline specification/types.md:91]: I had `Custom` here, but I looked at the rust SDK implementation. It was using this pattern already, and has `Other` which is nice.

File: specification/types.md (hunk 1 above)
Comment (toddbaert) [reply specification/types.md:81]: I like having a rust example. and this is clear to me.

**Review by toddbaert — APPROVED** [review]

File: specification/types.md [hunk 3]
```diff
@@ -72,6 +72,28 @@ A set of pre-defined reasons is enumerated below:
 | STALE           | The resolved value is non-authoritative or possibly out of date                                                                  |
 | ERROR           | The resolved value was the result of an error.                                                                                   |
 
+> [!NOTE]
+> The `reason` should not be limited to the reasons enumerated above. It can be any of the pre-defined reasons, or
+> any string value. Some type systems have features which can increase the ergonomics of `reason`, for instance a union
+of pre-defined types with a string, or a rust-style enumeration which allows for enumerated values to have associated
+content.
```
Comment (beeme1mr) [inline specification/types.md:79]:
```suggestion
> of pre-defined types with a string, or a rust-style enumeration which allows for enumerated values to have associated
> content.
```

**Review by beeme1mr — APPROVED** [review]

**Review by thomaspoignant — APPROVED** [review]

**Pull request merged**